contracts/src/PBHVerifier.sol

```diff
@@ -1,32 +1,154 @@
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @dev Whether a nullifier hash has been used already. Used to guarantee an action is only performed once by a single person
+    mapping(uint256 => bool) public nullifierHashes;
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                                  Functions                             ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @param sender The sender of this particular transaction or UserOp.
+    /// @param nonce Transaction/UserOp nonce.
+    /// @param callData Transaction/UserOp call data.
+    /// @param pbhPayload The PBH payload containing the proof data.
+    function verifyPbhProof(address sender, uint256 nonce, bytes calldata callData, PBHPayload memory pbhPayload)
+        public
+        virtual
+        onlyInitialized
+        onlyProxy
+    {
+        // First, we make sure this nullifier has not been used before.
+        if (nullifierHashes[pbhPayload.nullifierHash]) revert InvalidNullifier();
+
+        // Verify the external nullifier
+        PBHExternalNullifier.verify(pbhPayload.pbhExternalNullifier, numPbhPerMonth);
+
+        // If worldId address is set, proceed with on chain verification,
+        // otherwise assume verification has been done off chain by the builder.
+        if (address(_worldId) != address(0)) {
+            // We now generate the signal hash from the sender, nonce, and calldata
+            uint256 signalHash = abi.encodePacked(sender, nonce, callData).hashToField();
+
+            // We now verify the provided proof is valid and the user is verified by World ID
+            _worldId.verifyProof(
+                pbhPayload.root,
+                _GROUP_ID,
+                signalHash,
+                pbhPayload.nullifierHash,
+                pbhPayload.pbhExternalNullifier,
+                pbhPayload.proof
+            );
+        }
+
+        // We now record the user has done this, so they can't do it again (proof of uniqueness)
+        nullifierHashes[pbhPayload.nullifierHash] = true;
+
+        emit PBH(sender, nonce, callData, pbhPayload);
+    }
+
+    /// @notice Sets the number of PBH transactions allowed per month.
+    /// @param _numPbhPerMonth The number of allowed PBH transactions per month.
+    function setNumPbhPerMonth(uint8 _numPbhPerMonth) external virtual onlyOwner onlyProxy onlyInitialized {
+        numPbhPerMonth = _numPbhPerMonth;
+    }
+
+    /// @dev If the World ID address is set to 0, then it is assumed that verification will take place off chain.
+    /// @notice Sets the World ID instance that will be used for verifying proofs.
+    /// @param __worldId The World ID instance that will be used for verifying proofs.
+    function setWorldId(address __worldId) external virtual onlyOwner onlyProxy onlyInitialized {
+        _worldId = IWorldIDGroups(__worldId);
```

Comment:
We shold emit an event here if we are updating a state variable. Also lets remove the double underscore fro mthe arg.